### ISSUE TITLE
wb_add_data is very slow when there are non-adjacent columns of the POSIXct class.

### ISSUE THREAD

**Issue by apalacio9502** [JanMarvin/openxlsx2#1066]
Hi,

I was testing version 1.7 and noticed that wb_add_data is very slow when there are non-adjacent columns of the POSIXct class, which was not the case in version 1.6.

I have tested this on both Windows and Linux, and the behavior is consistent.

Example 1: when columns of the POSIXct class are adjacent.

```
wb <- openxlsx2::wb_workbook()
data <- data.frame(
  T1 = rep("test",5000),
  T2 = rep("test",5000),
  T3 = rep(as.POSIXct("2024-01-01"),5000),
  T4 = rep(as.POSIXct("2024-01-01"),5000)
)
system.time(
  wb <- wb %>% openxlsx2::wb_add_worksheet(
    sheet = "test",
    grid_lines = FALSE
  ) %>%
    openxlsx2::wb_add_data(
      sheet = "test",
      x = data,
      na.strings = ""
    )
)
   user  system elapsed 
  0.322   0.022   0.346 
```

Example 2: when columns of the POSIXct class are not adjacent.

```
wb <- openxlsx2::wb_workbook()
data <- data.frame(
  T1 = rep("test",5000),
  T3 = rep(as.POSIXct("2024-01-01"),5000),
  T2 = rep("test",5000),
  T4 = rep(as.POSIXct("2024-01-01"),5000)
)
system.time(
  wb <- wb %>% openxlsx2::wb_add_worksheet(
    sheet = "test",
    grid_lines = FALSE
  ) %>%
    openxlsx2::wb_add_data(
      sheet = "test",
      x = data,
      na.strings = ""
    )
)
   user  system elapsed 
  9.556   0.008   9.589 
```
As you can see when the columns are not adjacent it is 27 times slower than when they are adjacent

The current version of R that I am using is 4.4.1.

Regards,

**Comment by JanMarvin** [comment]
Thanks for the report @apalacio9502 , I have added a PR in #1067

But please note, I believe that I work through issues quickly enough and even though I don't like those "if I shovel in huge amounts of data, it isn't fast enough" issues at all, I still look at them and spend some time to work on them. That's why something like "you urgently need to look at this before the next release" feels a bit like pushing. And I don't like being pushed.

I looked at it, yes it was slow. Was it bad? Yes. Was it unbearably bad? No.


``` r
library(openxlsx2)

data1 <- data.frame(
  T1 = rep("test",5000),
  T2 = rep("test",5000),
  T3 = rep(as.POSIXct("2024-01-01"),5000),
  T4 = rep(as.POSIXct("2024-01-01"),5000)
)

data2 <- data.frame(
  T1 = rep("test",5000),
  T3 = rep(as.POSIXct("2024-01-01"),5000),
  T2 = rep("test",5000),
  T4 = rep(as.POSIXct("2024-01-01"),5000)
)

fun <- function(x) {
  wb_workbook()$
    add_worksheet(
      sheet = "test",
      grid_lines = FALSE
    )$
    add_data(
      sheet = "test",
      x = x,
      na.strings = ""
    )
}

res <- microbenchmark::microbenchmark(
  fun(data1),
  fun(data2),
  times = 5
)
#> Warning in microbenchmark::microbenchmark(fun(data1), fun(data2), times = 5):
#> less accurate nanosecond times to avoid potential integer overflows
```

``` r
res
#> Unit: milliseconds
#>        expr      min       lq     mean   median       uq       max neval
#>  fun(data1) 35.25594 35.43946 37.42409 38.32905 38.59879  39.49719     5
#>  fun(data2) 43.81309 46.57981 69.01942 46.99030 49.08442 158.62949     5
```

**Comment by apalacio9502** [comment]
Hi @JanMarvin,

Thank you for your response and quick solution.

I made the comment "I think it would be important to review this bug before the 1.8 release" because I believe this library has the potential to replace openxlsx in the future. This problem affects many users because dates are very common in data, and such problems hinder the adoption of the package.

I think I misunderstood the objective for which you use the roadmap and I apologize.

Regards,

**Comment by JanMarvin** [comment]
No reason to apologize, you were perfectly right to point out the issue.

I did some further testing while working on the issue and it obviously scales (small data set, low impact, big data set huge impact), which is ... bad to be honest and obviously something I didn't realize, when I initially coded it like this.
Even after the fix it is far from ideal. The entire issue stems from my attempt to fix #1023. But my solution was picking the very flexible way to apply cell styles. This delays things a bit, because initially we have to create a long vector consisting of all cells that are part of the dims range.

So this `"A1:A2,B1:B2"` will create this `"A1", "A2", "B1", "B2"`. In the slow code part we loop over a data frame, to fill it where needed with the cell information.

``` r
library(openxlsx2)
dims_to_dataframe("A1:A2,B1:B2", fill = TRUE)
#>    A  B
#> 1 A1 B1
#> 2 A2 B2
```

Whereas here we simply assume that each cell is required and fill it without additional checks (no construction of the long vector, no checking if cell is part of said vector):

```R
dims_to_dataframe("A1:B2", fill = TRUE)
```

Therefore the former is always slower than the latter which does not construct the cell vector. Since created it a bit confusing, the conversion from R character vector to C++ string vector happened in the looping over the cells.

But there should be more ways to speed things further up. Like checking initially if all cell ranges are equal sized and only check for required columns. But at least for now we have a better solution implemented.
